### PR TITLE
ADAPT-2517: Added autorelease github action.

### PR DESCRIPTION
**Pull request by nplowman**
# READY FOR REVIEW
- (Edit the above to reflect status)

# Summary
Adds autorelease Github action. This action:
- Automatically updates the changelog
- Increments the release version (both in package.json and with a git tag) based on Github label. (i.e. patch, minor, major, alpha, beta, rc)
- Runs 'npm run package' and commits any new changes to dist folder.
- Publishes latest package to npm.

# Needed By (Date)
- Next 2-3 days

# Urgency
- 3/10. (Process improvement, but nothing urgent or time sensitive).

# Steps to Test
This one will be a bit tricky to test, as the action is only fired when merging to "main". 
You may need to create a few dummy PRs that make changes to the README file or something along those lines to test the different scenarios. 

1. Create a new pull request, with base branch of "main". 
1. Add one of the following labels: "patch", "minor", "major", "alpha", "dev", or "rc"
1. Merge the PR.
1. Click the "Actions" tab in Github, validate that the action fired correctly and there are no log errors.
1. View the "Code" tab in Github. You should see that the CHANGELOG file has been updated with commit messages from the release, package.json updated with a new version number, and a new tag should also be visible.  

# Associated Issues and/or People
https://stanfordits.atlassian.net/browse/ADAPT-2517 

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

**Comment by nplowman** [comment]
A couple of other notes:
1. It looks like this repo does not yet have any tags. We should probably create an initial tag (like "v1.0.0") before merging this PR, as the script runs "git log $PREVIOUS_TAG..$NEW_TAG" to generate the changelog, and will likely fail if there is no previous tag.
2. We will also need to generate a publishing token for NPM, and add it to the Github secrets for this repo. The environment variable should be named "NPM_TOKEN"

File: .github/workflows/autorelease.yml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          branch: ${{ github.base_ref }}
+          tags: true
+          
+      # This step requires NPM_TOKEN to be set in environment variables.
+      - name: Publish to NPM
+        uses: JS-DevTools/npm-publish@v1
+        with:
+          token: ${{ secrets.NPM_TOKEN }}
+          access: "public"
```
Comment (nplowman) [inline .github/workflows/autorelease.yml:104]: @sherakama Let me know if we need to change the "access=public" flag. It looked like the package was public, but wasn't 100% sure.

File: .github/workflows/autorelease.yml [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        run: npm version prerelease --preid rc
+
+      # Automatically append new commit messages to changelog.
+      - name: "Update changelog"
+        run: |
+          NEW_TAG=$(git tag --sort=-taggerdate | head -1)
+          PREVIOUS_TAG=$(git tag --sort=-taggerdate | sed -n '2 p')
+          echo -e "$(git log --no-merges --pretty=format:%s --invert-grep --grep="Updated changelog with notes for" $PREVIOUS_TAG..$NEW_TAG)\n" >> CHANGELOG.txt
+          git add CHANGELOG.txt
+          git commit -m "Updated changelog with notes for $NEW_TAG"
```
Comment (sherakama) [inline .github/workflows/autorelease.yml:90]: Any reason to not use something like https://www.npmjs.com/package/auto-changelog?

Comment (nplowman) [reply .github/workflows/autorelease.yml:90]: @sherakama This wasn't on my radar, but I'm not opposed to giving it a try.

Comment (sherakama) [reply .github/workflows/autorelease.yml:90]: Give it a go then. It has nice formatting.

File: .github/workflows/autorelease.yml (hunk 1 above)
Comment (sherakama) [reply .github/workflows/autorelease.yml:105]: Yes, public is great.

**Comment by sherakama** [comment]
> It looks like this repo does not yet have any tags. We should probably create an initial tag (like "v1.0.0") before merging this PR, as the script runs "git log $PREVIOUS_TAG..$NEW_TAG" to generate the changelog, and will likely fail if there is no previous tag.

Yes, this is a good idea. We can do that.

**Comment by sherakama** [comment]
Ok, a 1.0.0 tag was punched.

File: .github/workflows/autorelease.yml [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+      # Run 'npm run package' and commit dist folder if there are any changes.
+      - name: npm run package
+        run: |
+          NEW_TAG=$(git tag --sort=-taggerdate | head -1)
+          npm run package
+          if [ $(git status | grep "dist\/") ]
+          then 
+            git add dist
+            git commit -m "Updated build package for release $NEW_TAG"
+          fi
```
Comment (sherakama) [inline .github/workflows/autorelease.yml:56]: Thanks for this. We often forget to run package.

File: .github/workflows/autorelease.yml (hunk 2 above)
Comment (nplowman) [reply .github/workflows/autorelease.yml]: @sherakama Action has been refactored to use auto changelog script.

File: package.json [hunk 4]
```diff
@@ -29,7 +29,8 @@
     "test": "jest",
     "test:watch": "jest --watch",
     "test:coverage": "jest --coverage",
-    "dev": "start-storybook -p 6006"
+    "dev": "start-storybook -p 6006",
+    "version": "auto-changelog -p && git add CHANGELOG.md"
```
Comment (nplowman) [inline package.json:33]: This ties the auto-changelog generation to the 'npm version' command. 
See https://www.npmjs.com/package/auto-changelog#what-you-might-do-if-youre-clever

Comment (sherakama) [reply package.json:33]: Smart.

**Review by sherakama — APPROVED** [review]
Ok, this looks good. Let's give it a whirl.

**Comment by sherakama** [comment]
@nplowman

Found a bug: https://github.com/SU-SWS/decanter-react/runs/2602767364?check_suite_focus=true

**Comment by nplowman** [comment]
> @nplowman
> 
> Found a bug: https://github.com/SU-SWS/decanter-react/runs/2602767364?check_suite_focus=true

Looks like the package-lock.json file is getting modified (maybe during the npm install step?)... I'll see if I can figure out what's going on there.

**Pull request merged**